### PR TITLE
[improve][pip] PIP-315: Configurable max delay limit for delayed delivery

### PR DESCRIPTION
**Pull request by KevinLiLu**
PIP-315: Configurable max delay limit for delayed delivery

### Motivation
Currently there is no max delay limit so a producer can specify any delay when publishing a message.

The purpose of this PIP is to introduce an optional configuration to limit the max allowed delay for delayed delivery.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Comment by KevinLiLu** [comment]
Hi @codelipenghui @jiazhai @merlimat @coderzc @wolfstudy @315157973, I saw you worked on delayed delivery code in the past. Would you mind providing some thoughts on this PIP? Thanks!

**Comment by KevinLiLu** [comment]
Hi @lhotari @tisonkun @mattisonchao @rdhabalia @poorbarcode @Technoboy- , could you review this PIP when you get a chance? Thanks!

File: pip/pip_315.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    PersistentTopic pTopic = (PersistentTopic) topic;
+    if (pTopic.isDelayedDeliveryEnabled()) {
+        headersAndPayload.markReaderIndex();
+        MessageMetadata msgMetadata = Commands.parseMessageMetadata(headersAndPayload);
+        headersAndPayload.resetReaderIndex();
+        if (msgMetadata.hasDeliverAtTime()) {
+            long deliverAtTime = msgMetadata.getDeliverAtTime();
+            long maxDeliveryDelayInSeconds = pTopic.getMaxDeliveryDelayInSeconds();
+            if (maxDeliveryDelayInSeconds > 0
+                    && deliverAtTime > System.currentTimeMillis() + maxDeliveryDelayInSeconds * 1000) {
```
Comment (315157973) [inline pip/pip_315.md:67]: This seems like a good proposal. I have some comments
1. Now the delay time accuracy reaches millisecond level, and the unit of the maximum delay threshold should also be consistent.
2. Each message will be judged, so we should try to reduce unnecessary repeated calculations, such as `maxDeliveryDelayInSeconds` * 1000

Comment (KevinLiLu) [reply pip/pip_315.md]: @315157973 Thanks for the review. You bring up a good point about the extra overhead per message when using seconds. I have updated the PIP to use milliseconds now.

File: pip/pip_315.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+```
+
+The max delivery delay check will occur in the broker's `Producer` class inside of `checkAndStartPublish` (same place as other checks such as `isEncryptionEnabled`).
+
+We will give a `ServerError.NotAllowedError` error if all of the following are true:
+1. Sending to a persistent topic
+2. Topic has `delayedDeliveryEnabled=true`
+3. `MessageMetadata` `deliver_at_time` has been specified
+4. Topic has `>0` value for `maxDeliveryDelayInMillis`
+5. `deliver_at_time` > `System.currentTimeMillis() + maxDeliveryDelayInMillis`
```
Comment (coderzc) [inline pip/pip_315.md:53]: Can we use `publishTime - deliverAtTime > maxDeliveryDelayInMillis` instead?

Comment (KevinLiLu) [reply pip/pip_315.md]: @coderzc Good idea. I have updated the PIP. Thanks for the suggestion!

**Review by coderzc — APPROVED** [review]

**Comment by 315157973** [comment]
This proposal needs to be voted in dev mail list at first

**Comment by KevinLiLu** [comment]
Hi @315157973 @coderzc, I have started the vote thread and added the link.

**Review by lhotari — APPROVED** [review]
Great work, @KevinLiLu! This is a very well written PIP. I'll follow up on the voting thread soon.

**Pull request merged**